File: 61-RotateList/61-RotateList.java
// Last updated: 7/31/2026, 8:29:01 PM
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode rotateRight(ListNode head, int k) {
        if (head == null) return head;
        ListNode temp = head;
        int len = 0;
        while (temp != null) {
            temp = temp.next;
            len++;
        }
        k %= len;
        if (k == 0) return head;
        temp = head;
        ListNode fast = head;
        for (int i = 0; i < k; i++) {
            fast = fast.next;
        }
        while (fast.next != null) {
            temp = temp.next;
            fast = fast.next;
        }
        fast.next = head;
        fast = temp.next;
        temp.next = null;
        return fast;
    }
}

File: 61-RotateList/61-RotateList.cpp
// Last updated: 8/9/2026, 1:06:02 PM
class Solution {
public:
    ListNode* rotateRight(ListNode* head, int k) {
        if (!head || !head->next || k == 0) return head;
        
        int len = 1;
        ListNode* temp = head;
        while (temp->next != nullptr) {
            len++;
            temp = temp->next;
        }

        k %= len;
        if (k == 0) return head;
        
        temp->next = head;
        for (int i = 0; i < len - k; i++) {
            temp = temp->next;
        }
        ListNode* newHead = temp->next;
        temp->next = nullptr;
        return newHead;
    }
};